src/ThingsBoard.h

```diff
@@ -217,52 +260,155 @@ class ThingsBoard
+      if (t.serializeKeyval(object) == false){
+        Logger::log("unable to serialize data");
+        return false;
+      }
+
+      if (measureJson(jsonBuffer) > PayloadSize - 1) {
+        Logger::log("too small buffer for JSON data");
+        return false;
+      }
       serializeJson(object, payload, sizeof(payload));
     }
     return telemetry ? sendTelemetryJson(payload) : sendAttributeJSON(payload);
   }
 
   // Processes RPC message
-  void process_message(char* topic, uint8_t* payload, unsigned int length);
+  void process_message(char* topic, uint8_t* payload, unsigned int length) {
+    RPC_Response r;
+    {
+      StaticJsonDocument<JSON_OBJECT_SIZE(MaxFieldsAmt)> jsonBuffer;
+      DeserializationError error = deserializeJson(jsonBuffer, payload, length);
+      if (error) {
+        Logger::log("unable to de-serialize RPC");
+        return;
+      }
+      const JsonObject &data = jsonBuffer.template as<JsonObject>();
+
+      const char *methodName = data["method"];
+
+      if (methodName) {
+        Logger::log("received RPC:");
+        Logger::log(methodName);
+      } else {
+        Logger::log("RPC method is NULL");
+        return;
+      }
+
+      for (size_t i = 0; i < sizeof(m_rpcCallbacks) / sizeof(*m_rpcCallbacks); ++i) {
+        if (m_rpcCallbacks[i].m_cb && !strcmp(m_rpcCallbacks[i].m_name, methodName)) {
+
+          Logger::log("calling RPC:");
+          Logger::log(methodName);
+
+          // Do not inform client, if parameter field is missing for some reason
+          if (!data.containsKey("params")) {
+            Logger::log("no parameters passed with RPC, passing null JSON");
+          }
+          // Getting non-existing field from JSON should automatically
+          // set JSONVariant to null
+          r = m_rpcCallbacks[i].m_cb(data["params"]);
+          break;
+        }
+      }
+    }
+    {
+      // Fill in response
+      char payload[PayloadSize] = {0};
+      StaticJsonDocument<JSON_OBJECT_SIZE(1)> respBuffer;
+      JsonVariant resp_obj = respBuffer.template to<JsonVariant>();
+
+      if (r.serializeKeyval(resp_obj) == false) {
+        Logger::log("unable to serialize data");
+        return false;
+      }
+
+      if (measureJson(respBuffer) > PayloadSize - 1) {
+        Logger::log("too small buffer for JSON data");
+        return false;
+      }
+      serializeJson(resp_obj, payload, sizeof(payload));
+
+      String responseTopic = String(topic);
+      responseTopic.replace("request", "response");
+      Logger::log("response:");
+      Logger::log(payload);
+      m_client.publish(responseTopic.c_str(), payload);
+    }
+  }
 
   // Sends array of attributes or telemetry to ThingsBoard
-  bool sendDataArray(const Telemetry *data, size_t data_count, bool telemetry = true);
+  bool sendDataArray(const Telemetry *data, size_t data_count, bool telemetry = true) {
+    if (MaxFieldsAmt < data_count) {
+      Logger::log("too small amount of JSON fields");
```

Comment:
I suggest to change it to `too much JSON fields passed` to be consistent with a documentation